Sources/Hyperspace/HTTP/HTTP.swift

```diff
@@ -127,53 +127,93 @@ public struct HTTP {
             self.init(data)
         }
     }
+
+    /// Represents an HTTP request
+    public struct Request: Equatable {
+
+        /// The `URL` to which the `Request` originated.
+        public let url: URL?
+
+        /// The HTTP header fields for this request.
+        public let headers: [String: String]?
+
+        /// The raw `Data` associated with the HTTP request, if any was provided.
+        public let body: Data?
+
+        /// Initialize a new `Request` with any given `URL`, HTTP headers and body.
+        /// - Parameters:
+        ///   - url: The `URL` for which this request was created.
+        ///   - data: The raw `Data` associated with the HTTP request, if any was provided.
+        ///   - headers: The HTTP header fields for this request.
+        public init(url: URL? = nil, headers: [String: String]? = nil, body: Data? = nil) {
+            self.url = url
+            self.headers = headers
+            self.body = body
+        }
+
+        /// Initialize a new `Request` given a URL request.
+        /// - Parameter urlRequest: The `URLRequest` instance used to initiate the request.
+        public init(urlRequest: URLRequest) {
+            self.init(url: urlRequest.url, headers: urlRequest.allHTTPHeaderFields, body: urlRequest.httpBody)
+        }
+    }
     
     /// Represents a HTTP response.
     public struct Response: Equatable {
 
-        /// The raw HTTP status code for this response.
+        /// The `HTTP.Request` object used to receive this response
+        public let request: Request
+
+        /// The raw HTTP status code for this response
         public let code: Int
-        
-        /// The URL from which the `Response` originated.
+
+        /// The `URL` from which the `Response` originated.
         public let url: URL?
```

Comment:
Since the `HTTP.Request` is part of the `HTTP.Response` now, should this just be a computed property that grabs it off of the request?

Reply:
The only reason it isn't now is in the case where you get redirected before received a response. I believe in that case you'd want to the `Request` and `Response` to have different `URL`s.

Reply:
Forgot about redirects, but that totally makes sense.